### PR TITLE
Introduce scalafmt-dynamic

### PR DESCRIPTION
**Pull request by tanishiking**
Close https://github.com/scalameta/sbt-scalafmt/issues/15

This PR introduces scalafmt-dynamic to sbt-scalafmt, enable it to switch scalafmt version using `version` setting, and removes `scalafmt-big` (or `scalafmt-core`) from dependency.

### Remove `scalafmtCli`, `scalafmtCliTest`, `scalafmtCliDiff`, and `scalafmtCliDiffTest`
Even though I released `scalafmt-cli 2.0.0-RC5` that can switch its running version using `version` setting, I'm on the fence whether we should still support `scalafmtCli` task :thinking: 

`scalafmt-cli` depends on both `scalafmt-dynamic` and `scalafmt-core`, and `scalafmt-cli#exceptionThrowingMain` runs the build version of scalafmt using `scalafmt-core` if `version` setting is missing or `version` is equal to its build version (in this case '2.0.0-RC5').
This means that if we depend on `scalafmt-cli`, `sbt-scalafmt` will depend on a specific version of `scalafmt-core` indirectly (in terms of `scalafmtCli*` tasks).

I believe that scalafmt integrations should loosely be coupled with a specific version of `scalafmt-core`. That's why, I unwillingly stop supporting `scalafmtCli*` tasks...

### `scalafmtConfig = taskKey[Option[File]]` to `scalafmtConfig = taskKey[File]`
If the file is missing, `sbt-scalafmt` will throw a exception.

Since `scalafmt-dynamic` requires `version` setting in `.scalafmt.conf`, now we shouldn't accept `None` for `scalafmtConfig`.
Indeed, we can run scalafmt without `version`, using `.withDefaultVersion(...)`, I prefer to avoid depending on a specific (default) version of scalafmt.

## Related issue
https://github.com/scalameta/scalafmt/issues/1318

### CODE REVIEW

File: plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala [hunk 1]
```diff
@@ -67,81 +61,64 @@ object ScalafmtPlugin extends AutoPlugin {
...
+      log: Logger,
+      writer: PrintWriter
   )(
       onError: (File, Throwable) => T,
       onFormat: (File, Input, Output) => T
   ): Seq[Option[T]] = {
+    val reporter = new ScalafmtSbtReporter(log, writer)
+    val scalafmtInstance = Scalafmt.create(this.getClass.getClassLoader).withReporter(reporter)
     sources
-      .withFilter(filterSource(_, config))
```
Comment (tanishiking) [inline plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala]: this filtering will be done by `scalafmt-dynamic`

File: plugin/src/sbt-test/scalafmt-sbt/sbt/build.sbt [hunk 2]
```diff
@@ -147,22 +158,22 @@ TaskKey[Unit]("check") := {
...
-      |object MainTest {
-      |  foo(
-      |    a, // comment
-      |    b
-      |  )
+      |object
+      |MainTest
+      |{
+      |  foo(a, // comment
+      |    b)
```
Comment (tanishiking) [inline plugin/src/sbt-test/scalafmt-sbt/sbt/build.sbt]: Previously, if `.scalafmt.conf` is missing, the default configuration is used for running scalafmt.
As of this commit, sbt-scalafmt does nothing if the configuration is missing.

File: plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala (hunk 1 above)
Comment (olafurpg) [reply plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala]: Nice!

File: plugin/src/sbt-test/scalafmt-sbt/sbt/build.sbt (hunk 2 above)
Comment (olafurpg) [reply plugin/src/sbt-test/scalafmt-sbt/sbt/build.sbt]: Good to test this 👍

**Comment by olafurpg** [comment]
@tanishiking what do you think about releasing RC5 after merging this PR?

**Comment by tanishiking** [comment]
Thank you for reviewing!

> I think it would be great if we can figure out how to continue supporting scalafmtCli down the road. This PR is already a huge improvements so I don't think this limitation is a blocking issue.
> The reflective API for the cli is args(Array[String]): Unit it might be possible to reflectively call it the same way scalafmt-dynamic does. I might be tempted to take a stab at this myself since I have considered started using sbt-scalafmt instead of the cli binary.

OK :+1: thanks!

> what do you think about releasing RC5 after merging this PR?

Sounds good!

**Comment by tanishiking** [comment]
Can you merge and release it whenever you have time?
(Or I'll do that if you give me a collaborator right on sbt-scalafmt ?)

**Comment by olafurpg** [comment]
I’m sorry I forgot to give you merge rights in this repo. You should have an invite now!

And yes, it would be good to update the organization name before releasing 👍

**Comment by tanishiking** [comment]
Wow, thanks! I'll cut the new release after changing the organization name.
Thanks again for the reviewing :)

**Pull request merged**